DataStructures/Lists/CircularBuffer.cs

```diff
@@ -0,0 +1,87 @@
+    public class CircularBuffer<T> {
+        /// <summary>
+        /// Initializes a circular buffer with initial length of 10
+        /// </summary>
+        public CircularBuffer(bool canOverride=true) : this(10,canOverride) {
+        }
+
+        /// <summary>
+        /// Initializes a circular buffer with given length
+        /// </summary>
+        /// <param name="length">The length of the buffer</param>
+        public CircularBuffer(int length, bool canOverride=true) {
+            if (length < 1) {
+                throw new ArgumentOutOfRangeException("length can not be zero or negative");
+            }
+            _circularBuffer = new T[length + 1];
+            _end = 0;
+            _start = 0;
+            CanOverride = canOverride;
+        }
+
+        /// <summary>
+        /// Writes value to the back of the buffer
+        /// </summary>
+        /// <param name="value">value to be added to the buffer</param>
+        public void Write(T value) {
+            if (CanOverride) {
+                InsertData(value);
+            }
+            else {
+                DontOverrides(value);
+            }
+        }
+
+        /// <summary>
+        /// Inserts data into the buffer when it is not filled up
+        /// </summary>
+        /// <param name="value"></param>
+        private void DontOverrides(T value) {
+            if (IsFilledUp) {
+                return;
+            }
+            InsertData(value);
+        }
+
+        /// <summary>
+        /// Inserts data into the buffer without checking if it is full
+        /// </summary>
+        /// <param name="value"></param>
+        private void InsertData(T value) {
+            _circularBuffer[_end] = value;
+            _end = (_end + 1) % _circularBuffer.Length;
+            if (_end == _start) {
+                _start = (_start + 1) % _circularBuffer.Length;
+            }
+        }
+
+        /// <summary>
+        ///     Reads value from the front of the buffer
+        ///     <summay>
+        ///         Use the IsEmpty property to know when to read appropriately.
+        ///     </summay>
+        /// </summary>
+        /// <returns></returns>
+        public T Read() {
+            var result = _circularBuffer[_start];
+            _circularBuffer[_start] = _circularBuffer[_end];
+            _start = (_start + 1) % _circularBuffer.Length;
+            return result;
+        }
+
+        public bool IsEmpty { get => _end == _start; }
+        public int Length { get => _circularBuffer.Length - 1; }
+        public bool IsFilledUp { get => ((_end + 1) % _circularBuffer.Length == _start) && !_circularBuffer[_start].Equals(_circularBuffer[_end]); }
+        /// <summary>
+        /// Controls whether data should be overridden when it is continously inserted without reading
+        /// </summary>
+        public bool CanOverride { get; }
+
+        private T[] _circularBuffer;
+        private int _end;
+        private int _start;
```

Comment:
I think more intuitive is put properties at begin of class, but it's not required to change, up to you.

Reply:
Since it's not part of our convention here, I will prefer to leave it that way. I have a personal arrangement for writing members of a class. Constructors, methods, events, properties and fields.